torch_datasets.py

```diff
@@ -0,0 +1,194 @@
+from typing import Dict, List, cast, Tuple
+import json
+import logging
+from overrides import overrides
+
+from torch.utils.data import Dataset as TorchDataset
+from torch.utils.data import DataLoader
+from torch.utils.data import Sampler, BatchSampler, SequentialSampler, SubsetRandomSampler
+
+from allennlp.common.registrable import Registrable
+from allennlp.common.util import add_noise_to_dict_values, lazy_groups_of
+from allennlp.data.batch import Batch as AllennlpBatch
+from allennlp.data.instance import Instance
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.data import Token
+from allennlp.common.file_utils import cached_path
+from allennlp.data.fields import Field, TextField, LabelField, MetadataField
+from allennlp.data.token_indexers import SingleIdTokenIndexer, TokenIndexer
+
+logger = logging.getLogger(__name__)
+
+class Dataset(TorchDataset, Registrable):
+    def __init__(self):
+        self.vocab: Vocabulary = None
+
+    def text_to_instance(self, *inputs) -> Instance:
+
+        raise NotImplementedError
+
+    def __getitem__(self) -> Instance:
+
+        raise NotImplementedError
+
+    def index_with(self, vocab: Vocabulary):
+        self.vocab = vocab
+
+
+"""
+Here we have two SNLI readers in both of the different styles.
+They are only slightly different.
+"""
+
+
+class SnliDataset(Dataset):
+    def __init__(
+        self, file_path: str, token_indexers: Dict[str, TokenIndexer] = None, lazy: bool = False
```

Comment:
As long as there's some other object that separates the `file_path` argument from the `token_indexers` argument, I'm ok with this.  The thing that's separated should be the one that's `Registrable`, though.